staging/src/k8s.io/client-go/discovery/cached/disk/round_tripper.go

```diff
@@ -63,3 +66,54 @@ func (rt *cacheRoundTripper) CancelRequest(req *http.Request) {
 }
 
 func (rt *cacheRoundTripper) WrappedRoundTripper() http.RoundTripper { return rt.rt.Transport }
+
+// A crcDiskCache is a cache backend for github.com/gregjones/httpcache. It is
+// similar to httpcache's diskcache package, but uses checksums to ensure cache
+// integrity rather than fsyncing each cache entry in order to avoid performance
+// degradation on MacOS.
+//
+// See https://github.com/kubernetes/kubernetes/issues/110753 for more.
+type crcDiskCache struct {
+	disk *diskv.Diskv
+}
+
+// Get the requested key from the cache on disk. If Get encounters an error, or
+// the returned value is not a CRC-32 checksum followed by bytes with a matching
+// checksum it will return false to indicate a cache miss.
+func (c *crcDiskCache) Get(key string) ([]byte, bool) {
+	b, err := c.disk.Read(sanitize(key))
+	if err != nil || len(b) < binary.MaxVarintLen32 {
+		return []byte{}, false
+	}
+
+	response := b[binary.MaxVarintLen32:]
+	sum, _ := binary.Uvarint(b[:binary.MaxVarintLen32])
+	if crc32.ChecksumIEEE(response) != uint32(sum) {
+		return []byte{}, false
+	}
+
+	return response, true
+}
+
+// Set writes the response to a file on disk. The filename will be the FNV-32a
+// hash of the key. The file will contain the CRC-32 checksum of the response
+// bytes, followed by said response bytes.
+func (c *crcDiskCache) Set(key string, response []byte) {
+	sum := make([]byte, binary.MaxVarintLen32)
+	_ = binary.PutUvarint(sum, uint64(crc32.ChecksumIEEE(response)))
+	_ = c.disk.Write(sanitize(key), append(sum, response...)) // Nothing we can do with this error.
+}
+
+func (c *crcDiskCache) Delete(key string) {
+	_ = c.disk.Erase(sanitize(key)) // Nothing we can do with this error.
+}
+
+// Sanitize an httpcache key such that it can be used as a diskv key, which must
+// be a valid filename. The httpcache key will either be the requested URL (if
+// the request method was GET) or "<method> <url>" for other methods, per the
+// httpcache.cacheKey function.
+func sanitize(key string) string {
+	h := fnv.New32a()
+	_, _ = h.Write([]byte(key)) // Writing to a hash never returns an error.
+	return fmt.Sprintf("%X", h.Sum32())
+}
```

Comment:
The reason to move away from MD5 is that it's broken. I think there is not really a reason to avoid cryptographic hashes? The worst thing that happens is maybe you accidentally prevent an attack that none of us have thought of?

Reply:
I don't feel strongly here, but using a cryptographic hash for non-sensitive data just in case feels a bit odd to me. The reason we hash here is simply to turn a request method and URL into a valid filename. The content of the file isn't hashed, so using a cryptographic hash would only serve to obfuscate what URL the content came from at the expense of a bit of speed.

Reply:
Yeah not sure I understand why we would want to use a cryptographic hash algo here. Maybe lower chance of collision but still seems like a bit much.

Reply:
Cryptographic hashes get as close as possible to a guarantee there's no bad values, or e.g. in this use, chunks of data that could be deleted without changing the sum. They are really not much more expensive than bad hash functions. Even though this cache is not that important, I think the extra computation is more than paid for by the extra guarantees.

Second point, I find "security mindset" is best understood as not creating *ingredients* for attacks. Non-cryptographic hashes are an attack ingredient. The remaining ingredients would perhaps be an unexpected use of this cache, and unexpected control of contents by attacker.

Reply:
Pushed a commit that switches to sha256.